From 199b63ea8c165f6ff989efe884505b21816a7783 Mon Sep 17 00:00:00 2001
From: Dat Vo <dat.vo.yb@renesas.com>
Date: Thu, 1 Jun 2023 18:05:01 +0700
Subject: [PATCH 2/5] PCIe/MSI support

ITS part is left untouched as ITS is handled by Xen.
Add patch for S4Sk

Signed-off-by: Yoshihiro Shimoda <yoshihiro.shimoda.uh@renesas.com>
Signed-off-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>
Signed-off-by: Dat Vo <dat.vo.yb@renesas.com>
---
 arch/arm64/boot/dts/renesas/r8a779f0-s4sk.dts |  2 +-
 .../boot/dts/renesas/r8a779f0-spider.dts      |  2 +-
 arch/arm64/boot/dts/renesas/r8a779f0.dtsi     | 21 ++++++++++++++-----
 arch/arm64/configs/defconfig                  |  1 +
 drivers/misc/pci_endpoint_test.c              |  6 ++++++
 .../pci/controller/dwc/pcie-designware-host.c | 10 +++++++++
 drivers/pci/controller/dwc/pcie-renesas.c     | 16 ++++++++++++++
 7 files changed, 51 insertions(+), 7 deletions(-)

diff --git a/arch/arm64/boot/dts/renesas/r8a779f0-s4sk.dts b/arch/arm64/boot/dts/renesas/r8a779f0-s4sk.dts
index b6c61a20cd0d..7f3645257022 100644
--- a/arch/arm64/boot/dts/renesas/r8a779f0-s4sk.dts
+++ b/arch/arm64/boot/dts/renesas/r8a779f0-s4sk.dts
@@ -255,7 +255,7 @@ &pcie_bus_clk {
 
 &pciec0 {
 	status = "okay";
-	pinctrl-0 = <&pcie0_pins>;
+	pinctrl-0 = <&pcie0_pins>, <&pcie1_pins>;
 	pinctrl-names = "default";
 	clkreq-gpios = <&gpio2 15 GPIO_ACTIVE_LOW>;
 };
diff --git a/arch/arm64/boot/dts/renesas/r8a779f0-spider.dts b/arch/arm64/boot/dts/renesas/r8a779f0-spider.dts
index 3bdbd2829501..137740aebd74 100644
--- a/arch/arm64/boot/dts/renesas/r8a779f0-spider.dts
+++ b/arch/arm64/boot/dts/renesas/r8a779f0-spider.dts
@@ -138,7 +138,7 @@ &pcie_bus_clk {
 
 &pciec0 {
 	status = "okay";
-	pinctrl-0 = <&pcie0_pins>;
+	pinctrl-0 = <&pcie0_pins>, <&pcie1_pins>;
 	pinctrl-names = "default";
 	clkreq-gpios = <&gpio2 15 GPIO_ACTIVE_LOW>;
 };
diff --git a/arch/arm64/boot/dts/renesas/r8a779f0.dtsi b/arch/arm64/boot/dts/renesas/r8a779f0.dtsi
index 3307f0f6577d..449e7e561245 100644
--- a/arch/arm64/boot/dts/renesas/r8a779f0.dtsi
+++ b/arch/arm64/boot/dts/renesas/r8a779f0.dtsi
@@ -1069,11 +1069,20 @@ ipmmu_mm: iommu@eefc0000 {
 		gic: interrupt-controller@f1000000 {
 			compatible = "arm,gic-v3";
 			#interrupt-cells = <3>;
-			#address-cells = <0>;
+			#address-cells = <2>;
+			#size-cells = <2>;
+			ranges;
 			interrupt-controller;
 			reg = <0x0 0xf1000000 0 0x20000>,
 			      <0x0 0xf1060000 0 0x110000>;
 			interrupts = <GIC_PPI 9 IRQ_TYPE_LEVEL_HIGH>;
+
+			gic_its: msi-controller@f1040000 {
+				compatible = "arm,gic-v3-its";
+				msi-controller;
+				#msi-cells = <1>;
+				reg = <0x0 0xf1040000 0 0x20000>;
+			};
 		};
 
 		prr: chipid@fff00044 {
@@ -1234,7 +1243,7 @@ pciec0: pcie@e65d0000 {
 			      <0 0xe65d5000 0 0x1200>,
 			      <0 0xe65d6200 0 0x0e00>,
 			      <0 0xe65d7000 0 0x1000>,
-			      <0 0xfe000000 0 0x10000>;
+			      <0 0xfe000000 0 0x400000>;
 			reg-names = "dbi",
 				    "atu",
 				    "dma",
@@ -1246,7 +1255,7 @@ pciec0: pcie@e65d0000 {
 			bus-range = <0x00 0xff>;
 			device_type = "pci";
 				 /* downstream IO */
-			ranges =  <0x81000000 0 0x00000000 0 0xfe000000 0 0x00010000
+			ranges =  <0x81000000 0 0x00000000 0 0xfe000000 0 0x00400000
 				 /* non-prefetchable memory */
 				  0x82000000 0 0x30000000 0 0x30000000 0 0x10000000>;
 				     /* Map all possible DDR as inbound ranges */
@@ -1261,6 +1270,7 @@ pciec0: pcie@e65d0000 {
 			interrupt-names = "msi", "dma", "err", "fatal",
 					  "nonfatal", "lp", "vndmsg";
 			#interrupt-cells = <1>;
+			msi-map = <0x0 &gic_its 0x0 0x1000>;
 			interrupt-map-mask = <0 0 0 7>;
 			interrupt-map = <0 0 0 1 &gic GIC_SPI 416 IRQ_TYPE_LEVEL_HIGH
 					 0 0 0 2 &gic GIC_SPI 416 IRQ_TYPE_LEVEL_HIGH
@@ -1317,7 +1327,7 @@ pciec1: pcie@e65d8000 {
 			      <0 0xe65dd000 0 0x1200>,
 			      <0 0xe65de200 0 0x0e00>,
 			      <0 0xe65df000 0 0x1000>,
-			      <0 0xee900000 0 0x10000>,
+			      <0 0xee900000 0 0x400000>,
 			      <0 0xe65d6200 0 0x0e00>;
 			reg-names = "dbi",
 				    "atu",
@@ -1331,7 +1341,7 @@ pciec1: pcie@e65d8000 {
 			bus-range = <0x00 0xff>;
 			device_type = "pci";
 				 /* downstream IO */
-			ranges = <0x81000000 0 0x00000000 0 0xee900000 0 0x00010000
+			ranges = <0x81000000 0 0x00000000 0 0xee900000 0 0x00400000
 				 /* non-prefetchable memory */
 				 0x82000000 0 0xc0000000 0 0xc0000000 0 0x10000000>;
 				     /* Map all possible DDR as inbound ranges */
@@ -1346,6 +1356,7 @@ pciec1: pcie@e65d8000 {
 			interrupt-names = "msi", "dma", "err", "fatal",
 					  "nonfatal", "lp", "vndmsg";
 			#interrupt-cells = <1>;
+			msi-map = <0x0 &gic_its 0x0 0x1000>;
 			interrupt-map-mask = <0 0 0 7>;
 			interrupt-map = <0 0 0 1 &gic GIC_SPI 423 IRQ_TYPE_LEVEL_HIGH
 					 0 0 0 2 &gic GIC_SPI 423 IRQ_TYPE_LEVEL_HIGH
diff --git a/arch/arm64/configs/defconfig b/arch/arm64/configs/defconfig
index 36112389bf58..7cfce08938d7 100644
--- a/arch/arm64/configs/defconfig
+++ b/arch/arm64/configs/defconfig
@@ -253,6 +253,7 @@ CONFIG_PCIE_RENESAS_EP=n
 CONFIG_PCI_ENDPOINT=y
 CONFIG_PCI_ENDPOINT_CONFIGFS=y
 CONFIG_PCI_EPF_TEST=m
+CONFIG_PCI_MSI=y
 CONFIG_DEVTMPFS=y
 CONFIG_DEVTMPFS_MOUNT=y
 CONFIG_FW_LOADER_USER_HELPER=y
diff --git a/drivers/misc/pci_endpoint_test.c b/drivers/misc/pci_endpoint_test.c
index 75c4f103d3ab..276dd411ee9b 100644
--- a/drivers/misc/pci_endpoint_test.c
+++ b/drivers/misc/pci_endpoint_test.c
@@ -825,11 +825,17 @@ static int pci_endpoint_test_probe(struct pci_dev *pdev,
 	init_completion(&test->irq_raised);
 	mutex_init(&test->mutex);
 
+/*
 	if ((dma_set_mask_and_coherent(&pdev->dev, DMA_BIT_MASK(48)) != 0) &&
 	    dma_set_mask_and_coherent(&pdev->dev, DMA_BIT_MASK(32)) != 0) {
 		dev_err(dev, "Cannot set DMA mask\n");
 		return -EINVAL;
 	}
+*/
+	if (dma_set_mask_and_coherent(&pdev->dev, DMA_BIT_MASK(32)) != 0) {
+		dev_err(dev, "Cannot set DMA mask\n");
+		return -EINVAL;
+	}
 
 	err = pci_enable_device(pdev);
 	if (err) {
diff --git a/drivers/pci/controller/dwc/pcie-designware-host.c b/drivers/pci/controller/dwc/pcie-designware-host.c
index 39f3b37d4033..fc6ac833c0a6 100644
--- a/drivers/pci/controller/dwc/pcie-designware-host.c
+++ b/drivers/pci/controller/dwc/pcie-designware-host.c
@@ -8,8 +8,11 @@
  * Author: Jingoo Han <jg1.han@samsung.com>
  */
 
+#define DEBUG
+
 #include <linux/irqchip/chained_irq.h>
 #include <linux/irqdomain.h>
+#include <linux/module.h>
 #include <linux/msi.h>
 #include <linux/of_address.h>
 #include <linux/of_pci.h>
@@ -19,6 +22,10 @@
 #include "../../pci.h"
 #include "pcie-designware.h"
 
+static int msi_init = 0;
+module_param(msi_init, int, 0644);
+MODULE_PARM_DESC(msi_init, "Use own msi_host_init");
+
 static struct pci_ops dw_pcie_ops;
 static struct pci_ops dw_child_pcie_ops;
 
@@ -277,6 +284,8 @@ static void dw_pcie_msi_init(struct dw_pcie_rp *pp)
 	if (!pci_msi_enabled() || !pp->has_msi_ctrl)
 		return;
 
+	printk("%s: msi_target = %llx\n", __func__, msi_target);
+
 	/* Program the msi_data */
 	dw_pcie_writel_dbi(pci, PCIE_MSI_ADDR_LO, lower_32_bits(msi_target));
 	dw_pcie_writel_dbi(pci, PCIE_MSI_ADDR_HI, upper_32_bits(msi_target));
@@ -459,6 +468,7 @@ int dw_pcie_host_init(struct dw_pcie_rp *pp)
 			goto err_deinit_host;
 		}
 
+		printk("%s: ======== msi_init = %d\n", __func__, msi_init);
 		if (pp->ops->msi_host_init) {
 			ret = pp->ops->msi_host_init(pp);
 			if (ret < 0)
diff --git a/drivers/pci/controller/dwc/pcie-renesas.c b/drivers/pci/controller/dwc/pcie-renesas.c
index 253083b7c58c..589c57834621 100644
--- a/drivers/pci/controller/dwc/pcie-renesas.c
+++ b/drivers/pci/controller/dwc/pcie-renesas.c
@@ -89,6 +89,18 @@
 #define CFG_SYS_ERR_RC          GENMASK(10, 9)
 #define CFG_SAFETY_UNCORR_CORR  GENMASK(5, 4)
 
+#define AXIINTCADDR		0x0a00	/* 0x6c00 */
+#define AXIINTCADDR_VAL		0xf1050040	/* FIXME */
+#define AXIINTCCONT		0x0a04	/* 0x6c04 */
+#define AXIINTCCONT_VAL		(BIT(31) | GENMASK(11, 2))
+
+/* PCIEC PHY */
+#define RCVRCTRLP0		0x0040
+#define  PHY0_RX1_TERM_ACDC	BIT(14)
+#define  PHY0_RX0_TERM_ACDC	BIT(13)
+
+#define REFCLKCTRLP0		0x0B8
+
 /* PCI Shadow offset */
 #define SHADOW_REG(x)		(0x2000 + (x))
 /* BAR Mask registers */
@@ -353,6 +365,10 @@ static void renesas_pcie_init_rc(struct renesas_pcie *pcie)
 	val |= CLK_REG | CLK_PM;
 	renesas_pcie_writel(pcie, PCIEPWRMNGCTRL, val);
 
+	printk("%s: ----------- set for INTC 64bit\n", __func__);
+	renesas_pcie_writel(pcie, AXIINTCADDR, AXIINTCADDR_VAL);
+	renesas_pcie_writel(pcie, AXIINTCCONT, AXIINTCCONT_VAL);
+
 	/* Enable DBI read-only registers for writing */
 	dw_pcie_dbi_ro_wr_en(pci);
 
-- 
2.43.0

